scripts/codemods/typed_flags.py

```diff
@@ -0,0 +1,182 @@
+        if_block: Optional[cst.If] = None
+        init: Optional[cst.FunctionDef] = None
+        body = list(node.body.body)
+        kwonly_params = [
+            cst.Param(cst.Name(flag_name), cst.Annotation(cst.Name("bool")), default=cst.Ellipsis())
+            for flag_name in all_flags
+        ]
+        for b in body.copy():
+            if m.matches(b, m.FunctionDef(m.Name("__init__"))) and isinstance(b, cst.FunctionDef):
+                deco_names = [getattr(deco.decorator, "value", "") for deco in b.decorators]
+                if "_generated" in deco_names:
+                    body.remove(b)
+                    continue
+                hide_behind_typechecking = False
+                init = b
+                break
+        node = node.with_deep_changes(node.body, body=body)
+
+        if hide_behind_typechecking:
+            # find the existing one if one exists
+            for b in body:
+                if m.matches(b, m.If(test=m.Name("TYPE_CHECKING"))):
+                    # iterate through the options to see if one of the body is __init__
+                    for line in b.body.body:  # type: ignore
+                        if m.matches(line, m.FunctionDef(m.Name("__init__"))):
+                            if_block = b  # type: ignore
+                            break
+                if if_block:
+                    break
+            else:
+                # one doesn't currently exist, great.
+                # so we have two options here... we could make a typechecking block from scratch or... we could cheat
+                code = textwrap.dedent(
+                    """
+                    if TYPE_CHECKING:
+                        @_generated
+                        def __init__(self):
+                            ...
+                    """
+                )
+                if_block = cst.parse_statement(code)  # type: ignore
+                assert if_block  # noqa: S101
+                # now we need to add this if_block into the CST of node
+                # find the first function definition and insert it before there
+                body = list(node.body.body)
+                for pos, b in enumerate(body):
+                    if m.matches(b, m.FunctionDef()):
+                        body.insert(pos, if_block)
+                        node = node.with_deep_changes(node.body, body=body)
+                        break
+                else:
+                    # so there wasn't any function, not sure why
+                    # doesn't really matter, we just insert at the end
+                    body.append(if_block)
+                    node = node.with_deep_changes(node.body, body=body)
+
+            # find the init
+            for b in if_block.body.body:
+                if m.matches(b, m.FunctionDef(m.Name("__init__"))) and isinstance(
+                    b, cst.FunctionDef
+                ):
+                    init = b
+                    break
```

Comment:
I think this could be moved up? In the loop it'd be just `init = line`, not sure about the `else` part but should be doable.

Reply:
what do you mean?

Reply:
```diff
diff --git a/scripts/codemods/typed_flags.py b/scripts/codemods/typed_flags.py
index cbdc1903..5dbcd1ec 100644
--- a/scripts/codemods/typed_flags.py
+++ b/scripts/codemods/typed_flags.py
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: MIT
 
 import textwrap
-from typing import List, Optional
+from typing import List, Optional, cast
 
 import libcst as cst
 import libcst.codemod.visitors as codevisitors
@@ -91,8 +91,10 @@ class FlagTypings(codemod.VisitorBasedCodemodCommand):
                     for line in b.body.body:  # type: ignore
                         if m.matches(line, m.FunctionDef(m.Name("__init__"))):
                             if_block = b  # type: ignore
+                            init = line
                             break
                 if if_block:
+                    assert init  # todo: could resolve this by getting rid of the nested loop
                     break
             else:
                 # one doesn't currently exist, great.
@@ -107,6 +109,7 @@ class FlagTypings(codemod.VisitorBasedCodemodCommand):
                 )
                 if_block = cst.parse_statement(code)  # type: ignore
                 assert if_block  # noqa: S101
+                init = cast("cst.FunctionDef", if_block.body.body[0])
                 # now we need to add this if_block into the CST of node
                 # find the first function definition and insert it before there
                 body = list(node.body.body)
@@ -121,15 +124,6 @@ class FlagTypings(codemod.VisitorBasedCodemodCommand):
                     body.append(if_block)
                     node = node.with_deep_changes(node.body, body=body)
 
-            # find the init
-            for b in if_block.body.body:
-                if m.matches(b, m.FunctionDef(m.Name("__init__"))) and isinstance(
-                    b, cst.FunctionDef
-                ):
-                    init = b
-                    break
-            else:
-                raise RuntimeError
             # add to the init the generated decorator
             old_init = init
             init = old_init.with_changes(decorators=[cst.Decorator(cst.Name("_generated"))])

```

Reply:
i don't think this is worth it tbh